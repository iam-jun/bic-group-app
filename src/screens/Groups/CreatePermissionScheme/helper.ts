import {IPermission, IRole, IScheme} from '~/interfaces/IGroup';
import {ROLE_TYPE} from '~/constants/permissionScheme';

export const getNewSchemeFromSystemScheme = (systemScheme: IScheme) => {
  let memberRoleIndex = 0;
  const roleKeys = ['scope', 'type', 'permissions', 'name'];
  const roles = systemScheme.roles?.map((role: IRole, i: number) => {
    if (role?.type === ROLE_TYPE.MEMBER) {
      memberRoleIndex = i;
    }
    return roleKeys.reduce((filtered, key) => {
      // @ts-ignore
      if (role[key]) filtered[key] = role[key];
      return filtered;
    }, {});
  });
  const newScheme: IScheme = {
    name: '',
    description: '',
    roles: roles as IRole[],
  };
  return {newScheme, memberRoleIndex};
};

export const getNewSchemeRolesOnUpdatePermission = (
  permission: IPermission,
  roleIndex: number,
  roles: IRole[],
) => {
  const newKey = permission?.key || '';
  const permissions = roles[roleIndex]?.permissions || [];
  if (permissions.includes?.(newKey)) {
    // key existed, should remove
    roles[roleIndex].permissions = permissions.filter(p => p !== newKey);
  } else {
    // add key
    permissions.push(newKey);
  }
  return [...roles];
};

export const getMemberRoleIndex = (schemeData: IScheme) => {
  let memberRoleIndex = 1;

  schemeData.roles?.map((role: IRole, i: number) => {
    if (role?.type === ROLE_TYPE.MEMBER) {
      memberRoleIndex = i;
      return;
    }
  });

  return memberRoleIndex;
};
