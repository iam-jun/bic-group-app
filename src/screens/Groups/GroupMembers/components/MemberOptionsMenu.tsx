import React from 'react';
import {View, StyleSheet} from 'react-native';
import i18next from 'i18next';
import {useDispatch} from 'react-redux';
import {useTheme} from 'react-native-paper';

import PrimaryItem from '~/beinComponents/list/items/PrimaryItem';
import BottomSheet from '~/beinComponents/BottomSheet';
import Text from '~/beinComponents/Text';
import Button from '~/beinComponents/Button';

import {IGroup, IGroupMemberRole, IGroupMembers} from '~/interfaces/IGroup';
import {useKeySelector} from '~/hooks/selector';
import groupsKeySelector from '../../redux/keySelector';
import {ITheme} from '~/theme/interfaces';
import useAuth from '~/hooks/auth';
import modalActions from '~/store/modal/actions';
import mainStack from '~/router/navigator/MainStack/stack';
import {useRootNavigation} from '~/hooks/navigation';
import groupsActions from '../../redux/actions';
import groupsDataHelper from '../../helper/GroupsDataHelper';

interface MemberOptionsMenuProps {
  groupId: number;
  modalizeRef: any;
  selectedMember?: IGroupMembers;
  onOptionsClosed: () => void;
}

const MemberOptionsMenu = ({
  groupId,
  modalizeRef,
  selectedMember,
  onOptionsClosed,
}: MemberOptionsMenuProps) => {
  const theme = useTheme() as ITheme;
  const styles = createStyle(theme);
  const dispatch = useDispatch();
  const {user} = useAuth();
  const {rootNavigation} = useRootNavigation();

  const can_manage_member = useKeySelector(
    groupsKeySelector.groupDetail.can_manage_member,
  );
  const can_setting = useKeySelector(groupsKeySelector.groupDetail.can_setting);

  const onPressMenuOption = (
    type:
      | 'view-profile'
      | 'send-message'
      | 'set-admin'
      | 'remove-admin'
      | 'remove-member'
      | 'leave-group',
  ) => {
    modalizeRef.current?.close();
    switch (type) {
      case 'view-profile':
        goToUserProfile(selectedMember);
        break;
      case 'set-admin':
        alertSettingAdmin(selectedMember);
        break;
      case 'remove-admin':
        onPressRemoveAdmin(selectedMember);
        break;
      case 'remove-member':
        alertRemovingMember(selectedMember);
        break;
      case 'leave-group':
        onPressLeave();
        break;
      default:
        dispatch(modalActions.showAlertNewFeature());
        break;
    }
  };

  const goToUserProfile = (selectedMember?: IGroupMembers) => {
    if (selectedMember) {
      rootNavigation.navigate(mainStack.userProfile, {
        userId: selectedMember.id,
      });
    }
  };

  const alertSettingAdmin = (selectedMember?: IGroupMembers) => {
    if (selectedMember) {
      const alertPayload = {
        iconName: 'Star',
        title: i18next.t('groups:modal_confirm_set_admin:title'),
        content: i18next.t('groups:modal_confirm_set_admin:description'),
        ContentComponent: Text.BodyS,
        cancelBtn: true,
        cancelBtnProps: {
          textColor: theme.colors.primary7,
          testID: 'member_options_menu.set_admin.cancel',
        },
        onConfirm: () => doSetAdmin(selectedMember),
        confirmLabel: i18next.t(
          'groups:modal_confirm_set_admin:button_confirm',
        ),
        ConfirmBtnComponent: Button.Secondary,
        confirmBtnProps: {
          highEmphasis: true,
          testID: 'member_options_menu.set_admin.confirm',
        },
      };
      alertPayload.content = alertPayload.content.replace(
        '{0}',
        `"${selectedMember?.fullname}"`,
      );
      dispatch(modalActions.showAlert(alertPayload));
    }
  };

  const doSetAdmin = (selectedMember: IGroupMembers) => {
    dispatch(
      groupsActions.setGroupAdmin({groupId, userIds: [selectedMember.id]}),
    );
  };

  const onPressRemoveAdmin = (selectedMember?: IGroupMembers) => {
    if (selectedMember) {
      // check if the current user is the last admin before revoking admin
      if (can_setting)
        return checkLastAdmin('remove', () =>
          alertRemovingAdmin(selectedMember),
        );
      alertRemovingAdmin(selectedMember);
    }
  };

  const alertRemovingAdmin = (selectedMember: IGroupMembers) => {
    const alertPayload = {
      iconName: 'StarHalfAlt',
      title: i18next.t('groups:modal_confirm_remove_admin:title'),
      content: i18next.t('groups:modal_confirm_remove_admin:description'),
      ContentComponent: Text.BodyS,
      cancelBtn: true,
      cancelBtnProps: {
        textColor: theme.colors.primary7,
        testID: 'member_options_menu.remove_admin.cancel',
      },
      onConfirm: () => doRemoveAdmin(selectedMember),
      confirmLabel: i18next.t(
        'groups:modal_confirm_remove_admin:button_confirm',
      ),
      confirmBtnProps: {
        testID: 'member_options_menu.remove_admin.confirm',
      },
      ConfirmBtnComponent: Button.Danger,
    };
    alertPayload.content = alertPayload.content.replace(
      '{0}',
      `"${selectedMember?.fullname}"`,
    );
    dispatch(modalActions.showAlert(alertPayload));
  };

  const doRemoveAdmin = (selectedMember: IGroupMembers) => {
    dispatch(
      groupsActions.removeGroupAdmin({groupId, userId: selectedMember?.id}),
    );
  };

  const onPressMemberButton = () => {
    dispatch(modalActions.clearToastMessage());
  };

  const checkLastAdmin = (type: string, callback: () => void) => {
    groupsDataHelper
      .getGroupMembers(groupId, {offset: 0, limit: 1})
      .then(data => {
        const adminCount = data?.GROUP_ADMIN?.user_count;
        if (adminCount > 1) {
          callback();
        } else {
          dispatch(
            modalActions.showHideToastMessage({
              content: `groups:error:last_admin_${type}`,
              props: {
                type: 'error',
                textProps: {useI18n: true},
                rightIcon: 'UsersAlt',
                rightText: 'Members',
                onPressRight: onPressMemberButton,
              },
              toastType: 'normal',
            }),
          );
        }
      })
      .catch(err => {
        console.error('[ERROR] error while fetching group members', err);
        dispatch(
          modalActions.showHideToastMessage({
            content: 'error:http:unknown',
            props: {textProps: {useI18n: true}, type: 'error'},
          }),
        );
      });
  };

  const alertRemovingMember = (selectedMember?: IGroupMembers) => {
    if (!selectedMember) {
      dispatch(
        modalActions.showHideToastMessage({
          content: 'No member selected',
          props: {type: 'error'},
        }),
      );
      return;
    }

    const {id: userId, fullname, username} = selectedMember;

    const content = i18next
      .t(`groups:modal_confirm_remove_member:final_alert`)
      .replace('{name}', `"${fullname}"`);

    const alertPayload = {
      iconName: 'RemoveUser',
      title: i18next.t('groups:modal_confirm_remove_member:title'),
      content: content,
      ContentComponent: Text.BodyS,
      cancelBtn: true,
      cancelBtnProps: {
        textColor: theme.colors.primary7,
        testID: 'member_options_menu.remove_member.cancel',
      },
      onConfirm: () => removeMember(userId, fullname),
      confirmLabel: i18next.t(
        'groups:modal_confirm_remove_member:button_remove',
      ),
      confirmBtnProps: {
        testID: 'member_options_menu.remove_member.confirm',
      },
      ConfirmBtnComponent: Button.Danger,
      children: null as React.ReactNode,
    };

    const renderAlertInnerGroups = (innerGroups: string[]) => {
      if (innerGroups.length === 0) return null;

      const count = innerGroups.length;
      let message = i18next
        .t('groups:modal_confirm_remove_member:alert_inner_groups')
        .replace('{0}', `${count}`);

      if (count === 1)
        message = message.replace(
          '1 other inner groups',
          'another inner group',
        );

      const first3groups = innerGroups.slice(0, 3);
      const groupsList = () => (
        <View style={styles.alertRemoveGroupsList}>
          {first3groups.map((name, index) => (
            <Text.BodyM key={index} style={styles.alertRemoveGroupsListItem}>
              â€¢ {name}
            </Text.BodyM>
          ))}
          {count > 3 && <Text.BodyS>...</Text.BodyS>}
        </View>
      );

      return (
        <>
          <Text.BodyS>{message}</Text.BodyS>
          {groupsList()}
        </>
      );
    };

    const getInnerGroupsNames = (innerGroups: any) => {
      const groupsRemovedFrom = [...innerGroups];

      if (groupsRemovedFrom.length === 0) {
        alertPayload.content = alertPayload.content.replace(
          '{other groups}',
          '',
        );
      } else {
        const otherGroups = groupsRemovedFromToString(groupsRemovedFrom);
        alertPayload.content = alertPayload.content.replace(
          '{other groups}',
          ` and ${otherGroups}`,
        );
        alertPayload.children = renderAlertInnerGroups(innerGroups);
      }

      dispatch(modalActions.showAlert(alertPayload));
    };

    handleLeaveInnerGroups(username, getInnerGroupsNames);
  };

  const removeMember = (userId: number, userFullname: string) => {
    dispatch(groupsActions.removeMember({groupId, userId, userFullname}));
  };

  const groupsRemovedFromToString = (groupList: string[]) => {
    if (groupList.length === 1) {
      return groupList[0];
    }

    return `${groupList.length} other inner groups: ${groupList.join(', ')}`;
  };

  const handleLeaveInnerGroups = (
    username: string,
    callback: (innerGroups: any) => void,
  ) => {
    // Get inner groups info (if any) when user leave/being removed from a group
    groupsDataHelper
      .getUserInnerGroups(groupId, username)
      .then(res => {
        const innerGroups = res.data.inner_groups.map(
          (group: IGroup) => group.name,
        );
        callback(innerGroups);
      })
      .catch(err => {
        console.error('Error while fetching user inner groups', err);
        dispatch(
          modalActions.showHideToastMessage({
            content: 'error:http:unknown',
            props: {textProps: {useI18n: true}, type: 'error'},
          }),
        );
      });
  };

  const onPressLeave = () => {
    // check if the current user is the last admin before leaving group
    if (can_setting) return checkLastAdmin('leave', alertLeaveGroup);
    alertLeaveGroup();
  };

  const alertLeaveGroup = () => {
    const alertPayload = {
      iconName: 'SignOutAlt',
      title: i18next.t('groups:modal_confirm_leave_group:title'),
      content: i18next.t('groups:modal_confirm_leave_group:description'),
      ContentComponent: Text.BodyS,
      cancelBtn: true,
      cancelBtnProps: {
        textColor: theme.colors.primary7,
      },
      onConfirm: () => doLeaveGroup(),
      confirmLabel: i18next.t('groups:modal_confirm_leave_group:button_leave'),
      ConfirmBtnComponent: Button.Danger,
    };

    const getInnerGroupsNames = (innerGroups: any) => {
      if (innerGroups.length > 0) {
        alertPayload.content =
          alertPayload.content +
          ` ${i18next.t(
            'groups:modal_confirm_leave_group:leave_inner_groups',
          )}`;

        const groupsLeaveToString = innerGroups.join(', ');
        alertPayload.content = alertPayload.content.replace(
          '{0}',
          groupsLeaveToString,
        );
      }

      dispatch(modalActions.showAlert(alertPayload));
    };

    handleLeaveInnerGroups(user.username, getInnerGroupsNames);
  };

  const doLeaveGroup = () => {
    dispatch(groupsActions.leaveGroup(groupId));
  };

  const isGroupAdmin = () => {
    return !!selectedMember?.roles?.find(
      (role: IGroupMemberRole) => role.type === 'GROUP_ADMIN',
    );
  };

  return (
    <BottomSheet
      modalizeRef={modalizeRef}
      onClosed={onOptionsClosed}
      ContentComponent={
        <View style={styles.bottomSheet}>
          <PrimaryItem
            testID="member_options_menu.view_profile"
            style={styles.menuOption}
            leftIcon={'UsersAlt'}
            leftIconProps={{icon: 'UsersAlt', size: 24}}
            title={i18next.t('groups:member_menu:label_view_profile')}
            onPress={() => onPressMenuOption('view-profile')}
          />
          {selectedMember?.username !== user?.username && (
            <PrimaryItem
              testID="member_options_menu.send_message"
              style={styles.menuOption}
              leftIcon={'iconSend'}
              leftIconProps={{icon: 'iconSend', size: 24}}
              title={i18next.t('groups:member_menu:label_send_message')}
              onPress={() => onPressMenuOption('send-message')}
            />
          )}
          {can_setting &&
            (isGroupAdmin() ? (
              <PrimaryItem
                testID="member_options_menu.remove_admin"
                style={styles.menuOption}
                leftIcon={'Star'}
                leftIconProps={{icon: 'Star', size: 24}}
                title={i18next.t('groups:member_menu:label_remove_as_admin')}
                onPress={() => onPressMenuOption('remove-admin')}
              />
            ) : (
              <PrimaryItem
                testID="member_options_menu.set_admin"
                style={styles.menuOption}
                leftIcon={'Star'}
                leftIconProps={{icon: 'Star', size: 24}}
                title={i18next.t('groups:member_menu:label_set_as_admin')}
                onPress={() => onPressMenuOption('set-admin')}
              />
            ))}
          {can_manage_member && selectedMember?.username !== user?.username && (
            <PrimaryItem
              testID="member_options_menu.remove_member"
              style={styles.menuOption}
              leftIcon={'UserTimes'}
              leftIconProps={{
                icon: 'UserTimes',
                size: 24,
                tintColor: theme.colors.error,
              }}
              title={i18next.t('groups:member_menu:label_remove_member')}
              titleProps={{color: theme.colors.error}}
              onPress={() => onPressMenuOption('remove-member')}
            />
          )}
          {selectedMember?.username === user?.username && (
            <PrimaryItem
              style={styles.menuOption}
              leftIcon={'SignOutAlt'}
              leftIconProps={{icon: 'SignOutAlt', size: 24}}
              title={i18next.t('groups:member_menu:label_leave_group')}
              onPress={() => onPressMenuOption('leave-group')}
            />
          )}
        </View>
      }
    />
  );
};

const createStyle = (theme: ITheme) => {
  const {spacing} = theme;
  return StyleSheet.create({
    bottomSheet: {
      paddingVertical: spacing.padding.tiny,
    },
    menuOption: {
      height: 44,
      paddingHorizontal: spacing.padding.large,
    },
    alertRemoveGroupsList: {
      marginBottom: spacing.margin.small,
    },
    alertRemoveGroupsListItem: {
      marginLeft: spacing.margin.small,
    },
  });
};

export default MemberOptionsMenu;
