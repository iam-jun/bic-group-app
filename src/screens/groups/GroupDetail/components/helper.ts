import useTimelineStore from '~/store/timeline';
import useGroupDetailStore from '../store';
import usePinContentStore from '~/components/PinContent/store';

export const onRefresh = async (props: { setIsRefreshing: (value: boolean) => void; groupId: string }) => {
  const { setIsRefreshing, groupId } = props;

  setIsRefreshing(true);

  await Promise.all([
    useGroupDetailStore.getState().actions.getGroupDetail({ groupId }),
    useTimelineStore.getState().actions.getPosts(groupId, true),
    usePinContentStore.getState().actions.getPinContentsGroup(groupId),
  ]);

  setIsRefreshing(false);
};
