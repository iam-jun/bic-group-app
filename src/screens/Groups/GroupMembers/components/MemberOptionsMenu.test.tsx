import React from 'react';

import {
  renderWithRedux,
  fireEvent,
  createTestStore,
  getHookReduxWrapper,
  renderHook,
  act,
} from '~/test/testUtils';
import MemberOptionsMenu from './MemberOptionsMenu';
import initialState from '~/store/initialState';
import {IGroupMembers} from '~/interfaces/IGroup';
import useRemoveMember from './useRemoveMember';
import * as navigationHook from '~/hooks/navigation';
import mainStack from '~/router/navigator/MainStack/stack';
import * as helper from '../../helper';
import modalActions from '~/store/modal/actions';
import i18next from 'i18next';

describe('MemberOptionsMenu component', () => {
  const baseSheetRef = jest.fn();
  const groupId = 1;
  const onOptionsClosed = jest.fn();
  const selectedMember = {};

  it('renders leave group option correctly', () => {
    const spy = jest.spyOn(helper, 'checkLastAdmin');

    const state = {...initialState};
    // @ts-ignore
    state.auth.user = {username: 'testname1'};
    const store = createTestStore(state);

    const selectedMember = {id: 1, username: 'testname1'};
    const {getByTestId} = renderWithRedux(
      <MemberOptionsMenu
        groupId={groupId}
        selectedMember={selectedMember}
        modalizeRef={baseSheetRef}
        onOptionsClosed={onOptionsClosed}
      />,
      store,
    );
    const itemComponent = getByTestId('member_options_menu.leave_group');
    expect(itemComponent).toBeDefined();
    expect(itemComponent.props).toHaveProperty('onClick');
    fireEvent.press(itemComponent);
    expect(spy).toBeCalled();
  });

  it('renders Remove member option correctly when admin clicks on another user', () => {
    const spy = jest.spyOn(helper, 'checkLastAdmin');

    const state = {...initialState};
    state.groups.groupDetail.can_manage_member = true;
    // @ts-ignore
    state.auth.user = {username: 'testname1'};
    const store = createTestStore(state);
    const selectedMember = {
      id: 1,
      username: 'testname2',
    } as IGroupMembers;

    const {getByTestId} = renderWithRedux(
      <MemberOptionsMenu
        groupId={groupId}
        modalizeRef={baseSheetRef}
        selectedMember={selectedMember}
        onOptionsClosed={onOptionsClosed}
      />,
      store,
    );
    const itemComponent = getByTestId('member_options_menu.remove_member');
    expect(itemComponent).toBeDefined();
    expect(itemComponent.props).toHaveProperty('onClick');
    fireEvent.press(itemComponent);
    expect(spy).toBeCalled();
  });

  it('should not render Remove member option correctly when admins click on themselves', () => {
    const state = {...initialState};
    state.groups.groupDetail.can_manage_member = true;
    // @ts-ignore
    state.auth.user = {username: 'testname1'};
    const store = createTestStore(state);
    const selectedMember = {
      id: 1,
      username: 'testname1',
    } as IGroupMembers;

    const {queryByTestId} = renderWithRedux(
      <MemberOptionsMenu
        groupId={groupId}
        modalizeRef={baseSheetRef}
        selectedMember={selectedMember}
        onOptionsClosed={onOptionsClosed}
      />,
      store,
    );
    const itemComponent = queryByTestId('member_options_menu.remove_member');
    expect(itemComponent).toBeNull();
  });

  it('getInnerGroupsNames should be done when there is no inner groups', () => {
    const renderInnerGroupsAlert = jest.fn();
    const state = {...initialState};
    const store = createTestStore(state);

    const wrapper = getHookReduxWrapper(store);
    const {result} = renderHook(
      () => useRemoveMember({groupId, selectedMember}),
      {
        wrapper,
      },
    );
    let pressResult;
    act(() => {
      pressResult = result.current.getInnerGroupsNames(
        [],
        renderInnerGroupsAlert,
      );
    });
    expect(pressResult).toBe(0);
  });

  it('getInnerGroupsNames should be done when there is 1 inner group', () => {
    const renderInnerGroupsAlert = jest.fn();
    const state = {...initialState};
    const store = createTestStore(state);

    const wrapper = getHookReduxWrapper(store);
    const {result} = renderHook(
      () => useRemoveMember({groupId, selectedMember}),
      {
        wrapper,
      },
    );
    let pressResult;
    act(() => {
      pressResult = result.current.getInnerGroupsNames(
        ['inner group 1'],
        renderInnerGroupsAlert,
      );
    });
    expect(pressResult).toBe(1);
  });

  it('getInnerGroupsNames should be done when there are at least 2 inner groups', () => {
    const renderInnerGroupsAlert = jest.fn();
    const state = {...initialState};
    const store = createTestStore(state);

    const wrapper = getHookReduxWrapper(store);
    const {result} = renderHook(
      () => useRemoveMember({groupId, selectedMember}),
      {
        wrapper,
      },
    );
    let pressResult;
    act(() => {
      pressResult = result.current.getInnerGroupsNames(
        ['inner group 1', 'inner group 2', 'inner group 3'],
        renderInnerGroupsAlert,
      );
    });
    expect(pressResult).toBe(1);
  });

  it('should render Remove admin option correctly', () => {
    const state = {...initialState};
    state.groups.groupDetail.can_setting = true;
    const store = createTestStore(state);

    const selectedMember = {
      id: 1,
      roles: [{type: 'GROUP_ADMIN'}],
    } as IGroupMembers;

    const {getByTestId} = renderWithRedux(
      <MemberOptionsMenu
        groupId={groupId}
        modalizeRef={baseSheetRef}
        selectedMember={selectedMember}
        onOptionsClosed={onOptionsClosed}
      />,
      store,
    );

    const item = getByTestId('member_options_menu.remove_admin');
    expect(item).toBeDefined();
  });

  it('should dispatch alertRemovingAdmin correctly', () => {
    const state = {...initialState};
    state.groups.groupDetail.can_setting = true;
    state.groups.groupMember = {
      // @ts-ignore
      group_admin: {user_count: 2},
    };
    const store = createTestStore(state);

    const selectedMember = {
      id: 1,
      roles: [{type: 'GROUP_ADMIN'}],
    } as IGroupMembers;

    const {getByTestId} = renderWithRedux(
      <MemberOptionsMenu
        groupId={groupId}
        modalizeRef={baseSheetRef}
        selectedMember={selectedMember}
        onOptionsClosed={onOptionsClosed}
      />,
      store,
    );

    const item = getByTestId('member_options_menu.remove_admin');
    expect(item).toBeDefined();
    expect(item.props).toHaveProperty('onClick');
  });

  it('should dispatch alert last admin error correctly', () => {
    const spy = jest.spyOn(modalActions, 'showHideToastMessage');

    const state = {...initialState};
    state.groups.groupDetail.can_setting = true;
    state.groups.groupMember = {group_admin: {user_count: 1}} as any;
    state.auth.user = {username: 'testname1'} as any;
    const store = createTestStore(state);

    const selectedMember = {
      id: 1,
      roles: [{type: 'GROUP_ADMIN'}],
    } as IGroupMembers;

    const {getByTestId} = renderWithRedux(
      <MemberOptionsMenu
        groupId={groupId}
        modalizeRef={baseSheetRef}
        selectedMember={selectedMember}
        onOptionsClosed={onOptionsClosed}
      />,
      store,
    );

    const item = getByTestId('member_options_menu.remove_admin');
    expect(item).toBeDefined();
    expect(item.props).toHaveProperty('onClick');
    fireEvent.press(item);
    expect(spy).toBeCalled();
  });

  it('should render set admin option correctly', () => {
    const spy = jest.spyOn(modalActions, 'showAlert');

    const state = {...initialState};
    state.groups.groupDetail.can_setting = true;
    state.auth.user = {username: 'testname1'} as any;
    const store = createTestStore(state);

    const selectedMember = {
      id: 1,
      roles: [{type: 'MEMBER'}],
    } as IGroupMembers;

    const {getByTestId} = renderWithRedux(
      <MemberOptionsMenu
        groupId={groupId}
        modalizeRef={baseSheetRef}
        selectedMember={selectedMember}
        onOptionsClosed={onOptionsClosed}
      />,
      store,
    );

    const item = getByTestId('member_options_menu.set_admin');
    expect(item).toBeDefined();
    expect(item.props).toHaveProperty('onClick');
    fireEvent.press(item);
    expect(spy).toBeCalledWith(
      expect.objectContaining({
        title: i18next.t('groups:modal_confirm_set_admin:title'),
      }),
    );
  });

  it('should navigate to user profile correctly when pressing View profile option', () => {
    const selectedMember = {
      id: 1,
      roles: [{type: 'MEMBER'}],
    } as IGroupMembers;

    const navigate = jest.fn();
    const rootNavigation = {navigate};
    jest.spyOn(navigationHook, 'useRootNavigation').mockImplementation(() => {
      return {rootNavigation} as any;
    });

    const {getByTestId} = renderWithRedux(
      <MemberOptionsMenu
        groupId={groupId}
        modalizeRef={baseSheetRef}
        selectedMember={selectedMember}
        onOptionsClosed={onOptionsClosed}
      />,
    );

    const item = getByTestId('member_options_menu.view_profile');
    expect(item).toBeDefined();
    fireEvent.press(item);
    expect(navigate).toBeCalledWith(mainStack.userProfile, {
      userId: selectedMember.id,
    });
  });

  it('renders Send message option correctly when admin clicks on another user', () => {
    const spy = jest.spyOn(modalActions, 'showAlertNewFeature');

    const state = {...initialState};
    // @ts-ignore
    state.auth.user = {username: 'testname1'};
    const store = createTestStore(state);
    const selectedMember = {
      id: 1,
      username: 'testname2',
    } as IGroupMembers;

    const {getByTestId} = renderWithRedux(
      <MemberOptionsMenu
        groupId={groupId}
        modalizeRef={baseSheetRef}
        selectedMember={selectedMember}
        onOptionsClosed={onOptionsClosed}
      />,
      store,
    );
    const itemComponent = getByTestId('member_options_menu.send_message');
    expect(itemComponent).toBeDefined();
    expect(itemComponent.props).toHaveProperty('onClick');
    fireEvent.press(itemComponent);
    expect(spy).toBeCalled();
  });
});
