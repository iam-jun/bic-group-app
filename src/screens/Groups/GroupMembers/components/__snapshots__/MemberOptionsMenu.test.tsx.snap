// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemberOptionsMenu component renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "paddingVertical": 4,
      }
    }
  >
    <View
      className="button"
    >
      <View
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 44,
            "opacity": 1,
            "paddingHorizontal": 16,
          }
        }
        testID="member_options_menu.view_profile"
      >
        <View
          accessibilityState={
            Object {
              "disabled": true,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          hitSlop={
            Object {
              "bottom": 10,
              "left": 10,
              "right": 10,
              "top": 10,
            }
          }
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                },
                Object {
                  "marginRight": 24,
                },
                Object {
                  "backgroundColor": undefined,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  undefined,
                  undefined,
                  undefined,
                ]
              }
            >
              <View
                className="icon-wrapper"
              >
                <RNSVGSvgView
                  align="xMidYMid"
                  bbHeight={24}
                  bbWidth={24}
                  fill="#2F3945"
                  focusable={false}
                  height={24}
                  meetOrSlice={0}
                  minX={0}
                  minY={0}
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "borderWidth": 0,
                      },
                      Object {},
                      Object {
                        "flex": 0,
                        "height": 24,
                        "width": 24,
                      },
                    ]
                  }
                  vbHeight={24}
                  vbWidth={24}
                  width={24}
                >
                  <RNSVGGroup
                    fill={4281284933}
                    propList={
                      Array [
                        "fill",
                      ]
                    }
                  >
                    <RNSVGPath
                      d="M12.3,12.22A4.92,4.92,0,0,0,14,8.5a5,5,0,0,0-10,0,4.92,4.92,0,0,0,1.7,3.72A8,8,0,0,0,1,19.5a1,1,0,0,0,2,0,6,6,0,0,1,12,0,1,1,0,0,0,2,0A8,8,0,0,0,12.3,12.22ZM9,11.5a3,3,0,1,1,3-3A3,3,0,0,1,9,11.5Zm9.74.32A5,5,0,0,0,15,3.5a1,1,0,0,0,0,2,3,3,0,0,1,3,3,3,3,0,0,1-1.5,2.59,1,1,0,0,0-.5.84,1,1,0,0,0,.45.86l.39.26.13.07a7,7,0,0,1,4,6.38,1,1,0,0,0,2,0A9,9,0,0,0,18.74,11.82Z"
                    />
                  </RNSVGGroup>
                </RNSVGSvgView>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <Text
            numberOfLines={2}
            style={
              Object {
                "color": "#2F3945",
                "fontFamily": "OpenSans-SemiBold",
                "fontSize": 14,
              }
            }
          >
            View profile
          </Text>
        </View>
      </View>
    </View>
    <View
      className="button"
    >
      <View
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 44,
            "opacity": 1,
            "paddingHorizontal": 16,
          }
        }
        testID="member_options_menu.leave_group"
      >
        <View
          accessibilityState={
            Object {
              "disabled": true,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          hitSlop={
            Object {
              "bottom": 10,
              "left": 10,
              "right": 10,
              "top": 10,
            }
          }
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                },
                Object {
                  "marginRight": 24,
                },
                Object {
                  "backgroundColor": undefined,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  undefined,
                  undefined,
                  undefined,
                ]
              }
            >
              <View
                className="icon-wrapper"
              >
                <RNSVGSvgView
                  align="xMidYMid"
                  bbHeight={24}
                  bbWidth={24}
                  fill="#2F3945"
                  focusable={false}
                  height={24}
                  meetOrSlice={0}
                  minX={0}
                  minY={0}
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "borderWidth": 0,
                      },
                      Object {},
                      Object {
                        "flex": 0,
                        "height": 24,
                        "width": 24,
                      },
                    ]
                  }
                  vbHeight={24}
                  vbWidth={24}
                  width={24}
                >
                  <RNSVGGroup
                    fill={4281284933}
                    propList={
                      Array [
                        "fill",
                      ]
                    }
                  >
                    <RNSVGPath
                      d="M12.59,13l-2.3,2.29a1,1,0,0,0,0,1.42,1,1,0,0,0,1.42,0l4-4a1,1,0,0,0,.21-.33,1,1,0,0,0,0-.76,1,1,0,0,0-.21-.33l-4-4a1,1,0,1,0-1.42,1.42L12.59,11H3a1,1,0,0,0,0,2ZM12,2A10,10,0,0,0,3,7.55a1,1,0,0,0,1.8.9A8,8,0,1,1,12,20a7.93,7.93,0,0,1-7.16-4.45,1,1,0,0,0-1.8.9A10,10,0,1,0,12,2Z"
                    />
                  </RNSVGGroup>
                </RNSVGSvgView>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <Text
            numberOfLines={2}
            style={
              Object {
                "color": "#2F3945",
                "fontFamily": "OpenSans-SemiBold",
                "fontSize": 14,
              }
            }
          >
            Leave Group
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
